kernel, refactor: return error status on all fatal errors

ryanofsky opened the pull request:
Return util::Result objects from all functions that can trigger fatal errors.

There are many validation functions that handle failures by calling `AbortNode` and triggering shutdowns, without returning error information to their callers. This makes error handling in libbitcoinkernel application code difficult, because the only way to handle these errors is to register for [notification callbacks](https://github.com/bitcoin/bitcoin/blob/master/src/kernel/notifications_interface.h). Improve this by making all functions that trigger fatal errors return util::Result objects with the error information.

This PR is a pure refactoring that returns extra result information from functions without changing their behavior. It's a possible alternative to and subset of #29642, which adds similar return information but also makes behavior changes and exposes a FatalError type.

---

<!-- begin based-on -->
**This is based on #25665.** The non-base commits are:

- [`4754af95750` refactor, kernel: Add FlushStatus / FlushResult types](https://github.com/bitcoin/bitcoin/pull/29700/commits/4754af95750b1debfcde3b03448e14b07f9236b2)
- [`fa892ba5080` refactor: Add InfoType field to util::Result](https://github.com/bitcoin/bitcoin/pull/29700/commits/fa892ba50808acf5ef6aa287eba88bb99d689a06)
- [`79db3562c96` refactor, blockstorage: Return FlushResult from flush methods](https://github.com/bitcoin/bitcoin/pull/29700/commits/79db3562c966a199468481c4b2e60af978e38b09)
- [`4b91a8e1422` refactor, blockstorage: Return fatal errors from block writes](https://github.com/bitcoin/bitcoin/pull/29700/commits/4b91a8e1422eefbf6c97d167b72374ccecf2ce84)
- [`01aab1e5d66` refactor, validation: Switch to result.Update](https://github.com/bitcoin/bitcoin/pull/29700/commits/01aab1e5d66033b6551416cd20429bfca498ce94)
- [`4704e208e13` refactor, blockstorage: Return fatal error from LoadBlockIndex](https://github.com/bitcoin/bitcoin/pull/29700/commits/4704e208e13a5f7772d5ed5274ac981d72856074)
- [`6dc2c1376a7` refactor, validation: Return fatal errors from FlushStateToDisk](https://github.com/bitcoin/bitcoin/pull/29700/commits/6dc2c1376a7e377709c698a59116f4b476e875e8)
- [`e0622fd7c03` refactor, validation: Return fatal errors from mempool accept functions](https://github.com/bitcoin/bitcoin/pull/29700/commits/e0622fd7c031c0ce6f1e27b69d5961b8a4604dca)
- [`01de9f6426f` refactor, validation: Return fatal errors from assumeutxo snapshot functions](https://github.com/bitcoin/bitcoin/pull/29700/commits/01de9f6426fd29ff207d1882c8739c9bef09106c)
- [`69ef13ead2c` refactor, validation: Return fatal errors from block connect functions](https://github.com/bitcoin/bitcoin/pull/29700/commits/69ef13ead2c165de872f1614f40beaa8490a0aaa)
- [`2999cf7382a` refactor, validation: Return fatal errors from activate best chain functions](https://github.com/bitcoin/bitcoin/pull/29700/commits/2999cf7382a6c12648c540851382f37802807b0b)
- [`36ede1f7d1d` refactor, validation: Return fatal errors from new block functions](https://github.com/bitcoin/bitcoin/pull/29700/commits/36ede1f7d1d5e3db6997446e493df751c023f007)
- [`719b09533f5` refactor, blockstorage: Return fatal error from ImportBlocks](https://github.com/bitcoin/bitcoin/pull/29700/commits/719b09533f50b37c97698b538b6c1a8690627741)
- [`6542bfe144b` refactor, validation: Return more errors from VerifyDB](https://github.com/bitcoin/bitcoin/pull/29700/commits/6542bfe144b1f735d628ce59973c37a55a48a7fc)<!-- end -->

ryanofsky committed:
refactor: Add util::Result::Update() method

Add util::Result Update method to make it possible to change the value of an
existing Result object. This will be useful for functions that can return
multiple error and warning messages, and want to be able to change the result
value while keeping existing messages.

ryanofsky committed:
refactor: Add util::Result failure values

Add util::Result support for returning more error information. For better error
handling, adds the ability to return a value on failure, not just a value on
success. This is a key missing feature that made the result class not useful
for functions like LoadChainstate() which produce different errors that need to
be handled differently.

This change also makes some more minor improvements:

- Smaller type size. On 64-bit platforms, util::Result<int> is 16 bytes, and
  util::Result<void> is 8 bytes. Previously util::Result<int> and
  util::Result<void> were 72 bytes.

- More documentation and tests.

ryanofsky committed:
test: add static test for util::Result memory usage

Suggested by Martin Leitner-Ankerl <martin.ankerl@gmail.com>
https://github.com/bitcoin/bitcoin/pull/25722#discussion_r1174298529

Co-authored-by: Martin Leitner-Ankerl <martin.ankerl@gmail.com>

ryanofsky committed:
refactor, validation: Return fatal errors from activate best chain functions

Return fatal errors from ActivateBestChain, ActivateBestChainStep, and
PreciousBlock functions.  Also add nodiscard annotations so callers handle the
result values. Two callers, ProcessNewBlock and LoadExternalBlockFile, will be
updated in an upcoming commits to bubble results up to its callers.

ryanofsky committed:
refactor, validation: Return fatal errors from new block functions

Return fatal errors from AcceptBlock, ProcessNewBlock, TestBlockValidity,
LoadGenesisBlock, and LoadExternalBlockFile. Also add nodiscard annotations so
callers handle the result values.

ryanofsky committed:
refactor, blockstorage: Return fatal error from LoadBlockIndex

Return fatal error and interrupt status from LoadBlockIndex functions and
update callers to use new result types.

ryanofsky committed:
refactor, kernel: Add FlushStatus / FlushResult types

ryanofsky committed:
refactor, validation: Return more errors from VerifyDB

Return ConnectBlock errors from VerifyDB.

ryanofsky committed:
refactor, validation: Return fatal errors from block connect functions

Return fatal errors from ConnectBlock, ConnectTip, DisconnectTip, and
InvalidateBlock. Also add nodiscard annotations so callers handle the result
values. Three callers: ActivateBestChainStep TestBlockValidity, and
CVerifyDB::VerifyDB will be updated in upcoming commits to bubble results up
to their callers.

ryanofsky committed:
refactor, validation: Return fatal errors from FlushStateToDisk

Return fatal errors from the Chainstate::FlushStateToDisk method and several
small, related methods which wrap it: ForceFlushStateToDisk, PruneAndFlush,
ResizeCoinsCaches, and MaybeRebalanceCaches.

Also add nodiscard annotations so callers do not accidentally ignore the result
values. Callers in init and rpc files are updated to explicitly ignore the
flush results, and other callers (AcceptToMemoryPool, ProcessNewPackage,
DisconnectTip, ConnectTip, ActivateBestChainStep, ActivateSnapshot,
MaybeCompleteSnapshotValidation) are updated to store the results in this
commit, and will be updated in upcoming commits to bubble results up to their
callers.

ryanofsky committed:
refactor, blockstorage: Return fatal error from ImportBlocks

Return fatal error ImportBlocks function and add nodiscard annotation.

ryanofsky committed:
refactor, validation: Return fatal errors from mempool accept functions

Return fatal errors from AcceptToMemoryPool ProcessNewPackage,
ProcessTransaction, MaybeUpdateMempoolForReorg, and LoadMempool functions.

Also add nodiscard annotations so callers handle the result values. Two callers
ActivateBestChainStep and InvalidateBlock will be updated in upcoming commits
to bubble results up to their callers.

ryanofsky committed:
refactor: Add InfoType field to util::Result

Add new InfoType field to util::Result which unlike SuccessType and FailureType
is useful for returning extra information in a result regardless of whether the
function succeeded for failed.

This will be used to return FlushStatus values from libbitcoinkernel functions
indicating if flushed data in addition to whether they succeeded or failed.

The InfoType field is stored directly in the util::Result object, unlike
FailureType and MessagesType which are stored in dynamically allocated memory
and require an extra memory allocation if they are accessed.

Graysonbarton reviewed (approved):
Fixing fatal errors.

ryanofsky committed:
refactor, blockstorage: Return FlushResult from flush methods

Return FlushResult instead of bool from BlockStorage FlushUndoFile,
FlushBlockFile, FlushChainstateBlockFile methods and update all callers of
these methods to use the FlushResult type internally and provide context
information for the flush failure. Three callers:
BlockManager::FindNextBlockPos, BlockManager::WriteUndoDataForBlock, and
Chainstate::FlushStateToDisk will be updated in upcoming commits to bubble
results up to their callers.

ryanofsky committed:
refactor: Add util::Result multiple error and warning messages

Add util::Result support for returning warning messages and multiple errors,
not just a single error string. This provides a way for functions to report
errors and warnings in a standard way, and simplifies interfaces.

The functionality is unit tested here, and put to use in followup PR
https://github.com/bitcoin/bitcoin/pull/25722

ryanofsky commented:
Rebased 4d2c9de24916f8d69514ea7c7251136e2762fa5c -> f65fa8c91130931713848a97606d5add0fc9b8c5 ([`pr/fatalresult.13`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.13) -> [`pr/fatalresult.14`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.13-rebase..pr/fatalresult.14)) due to silent conflict with #28970

ryanofsky committed:
refactor, validation: Switch to result.Update

Use result.Update in CompleteChainstateInit() and
ChainstateManager::ActivateSnapshot() so it is possible for them to return
warning messages (about flush failures) in upcoming commits.

CompleteChainstateInit() was previously changed to use util::Result in #25665
and ChainstateManager::ActivateSnapshot() was changed to use it in #30267, but
previously these functions only returned Result values themselves, and did not
call other functions that return Result values. Now, some functions they are
calling will also return Result values, so refactor these functions to use
result.Update so they can merge results and return complete error and warning
messages.

ryanofsky committed:
refactor, blockstorage: Return fatal errors from block writes

Return fatal errors from BlockManager methods that write block data. Also
update callers to use the new result types. The callers will be changed to
bubble up the results to their callers in subsequent commits.

ryanofsky committed:
refactor, validation: Return fatal errors from assumeutxo snapshot functions

Return fatal errors from ActivateSnapshot, MaybeCompleteSnapshotValidation, and
ValidatedSnapshotCleanup functions. Also add nodiscard annotations so callers
handle the result values. One caller, ConnectTip, will be updated in an
upcoming commit to bubble results up to its callers.

ryanofsky commented:
Rebased 692fc11f0aae3c8013f6f01d133139ce8eba7135 -> ad1e73590d102edf3738986c1382b5e36a0ed727 ([`pr/fatalresult.17`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.17) -> [`pr/fatalresult.18`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.17-rebase..pr/fatalresult.18)) to fix conflicts

ryanofsky commented:
Rebased ad1e73590d102edf3738986c1382b5e36a0ed727 -> e04e259146238f9a733edf3fd7b192db95f66e71 ([`pr/fatalresult.18`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.18) -> [`pr/fatalresult.19`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.19), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.18-rebase..pr/fatalresult.19)) to fix conflicts with #29668 and #30428

ryanofsky committed:
wallet: fix clang-tidy warning performance-no-automatic-move

Previous commit causes the warning to be triggered, but the suboptimal return
from const statement was added earlier in 517e204bacd9d.

Warning was causing CI failure https://cirrus-ci.com/task/6159914970644480

ryanofsky commented on src/node/blockstorage.cpp:
```diff
@@ -494,10 +502,11 @@ bool BlockManager::WriteBlockIndexDB()
 {
-    if (!LoadBlockIndex(snapshot_blockhash)) {
-        return false;
+    auto result{LoadBlockIndexData(snapshot_blockhash)};
+    if (!result || IsInterrupted(*result)) {
+        return result;
```
re: https://github.com/bitcoin/bitcoin/pull/29700#discussion_r1693183422

> Seems fragile to use `util::Result<InterruptResult>`. Wrapping one result into another is fragile, because call-sites may easily forget to unwrap the inner result, especially if the inner result is otherwise `void`. My recommendation would be to treat interruption as `util::Error`, so that callers don't miss it if they check for errors (but forget to check for interruption), but are free to check it, if they inspect the precise error.

It sounds like you want `Result<InterruptResult, void>` to be replaced with `Result<void, InterruptResult>` which would be ok with me. I don't think the difference between these things is very significant, but I agree maybe if the caller forgets to check for the interrupted status, incorrectly treating the interrupted status like a failure might be safer than incorrectly treating the interrupted status like a success.

ryanofsky reviewed (commented):
Rebased e22926a383223b286fe97a12ea4efaa81414bb41 -> 3ab9a46ac67964c9d5ad54120b680b5eee7f16f0 ([`pr/fatalresult.20`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.20) -> [`pr/fatalresult.21`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.21), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.20-rebase..pr/fatalresult.21)) due to conflicts with #30517 and #30497

TheCharlatan commented:
Can you rebase this?

ryanofsky committed:
refactor: Use util::Result class in LoadChainstate and VerifyLoadedChainstate

ryanofsky commented:
Rebased 3ab9a46ac67964c9d5ad54120b680b5eee7f16f0 -> 05438605ba95df14705bf2f57924455f3b2e40e1 ([`pr/fatalresult.21`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.21) -> [`pr/fatalresult.22`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.22), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.21-rebase..pr/fatalresult.22)) due to various conflicts. Still have not moved InterruptResult from success type to failure type yet (https://github.com/bitcoin/bitcoin/pull/29700#discussion_r1707623223), but would like to try that
Updated 05438605ba95df14705bf2f57924455f3b2e40e1 -> 37edd648608b35ee8647b24a2810cb68318d0e88 ([`pr/fatalresult.22`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.22) -> [`pr/fatalresult.23`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.23), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.22..pr/fatalresult.23)) to fix feature_assumeutxo.py errors https://github.com/bitcoin/bitcoin/actions/runs/11665121674/job/32477102140?pr=29700
Updated 37edd648608b35ee8647b24a2810cb68318d0e88 -> c8b555b18dc149f43f3fffdff7962c3e69e655a9 ([`pr/fatalresult.23`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.23) -> [`pr/fatalresult.24`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.24), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.23..pr/fatalresult.24)) to fix windows stderr line ending error https://github.com/bitcoin/bitcoin/actions/runs/11667772974/job/32485867701#step:12:131
Updated c8b555b18dc149f43f3fffdff7962c3e69e655a9 -> a1077c21dd8995c5f11a56b65432da759a3c1b8e ([`pr/fatalresult.24`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.24) -> [`pr/fatalresult.25`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.25), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.24..pr/fatalresult.25)) to fix windows errors caused by previous fix (https://github.com/bitcoin/bitcoin/actions/runs/11674634334/job/32507680468)
Rebased a1077c21dd8995c5f11a56b65432da759a3c1b8e -> af226a6a39ebab9d5886fcccdfbe63ea7da4fd9e ([`pr/fatalresult.25`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.25) -> [`pr/fatalresult.26`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.26), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.25-rebase..pr/fatalresult.26)) due to various conflicts
Rebased af226a6a39ebab9d5886fcccdfbe63ea7da4fd9e -> 63aa292c8e68c51cf49baba799cbb6aa573dec82 ([`pr/fatalresult.26`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.26) -> [`pr/fatalresult.27`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.27), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.26-rebase..pr/fatalresult.27)) due to conflict with #31072
Rebased 63aa292c8e68c51cf49baba799cbb6aa573dec82 -> 07d8ec809e10872e5bf3c870abab365bdcb9adba ([`pr/fatalresult.27`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.27) -> [`pr/fatalresult.28`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.28), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.27-rebase..pr/fatalresult.28)) due to conflicts with #31490, #30965, and #31439
Rebased 07d8ec809e10872e5bf3c870abab365bdcb9adba -> b795ab78b725c3ece4a5aca40330724e4929bc7f ([`pr/fatalresult.28`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.28) -> [`pr/fatalresult.29`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.29), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.28-rebase..pr/fatalresult.29)) due to conflict with #32033
Rebased b795ab78b725c3ece4a5aca40330724e4929bc7f -> 42a73b451cd68be443c7a0dd94f7a00477775e23 ([`pr/fatalresult.29`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.29) -> [`pr/fatalresult.30`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.30), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.29-rebase..pr/fatalresult.30)) due to silent conflict with #31910 causing nodiscard error https://cirrus-ci.com/task/5763662914256896
Rebased 42a73b451cd68be443c7a0dd94f7a00477775e23 -> f169e600c0868f96acf42bc3847acfa220db925d ([`pr/fatalresult.30`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.30) -> [`pr/fatalresult.31`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.31), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.30-rebase..pr/fatalresult.31)) due to various conflicts
Rebased f169e600c0868f96acf42bc3847acfa220db925d -> a4cf35ddbde791e491375217470649c0e844955b ([`pr/fatalresult.31`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.31) -> [`pr/fatalresult.32`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.32), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.31-rebase..pr/fatalresult.32)) due to conflict with #31385
<!-- begin push-33 -->
Updated a4cf35ddbde791e491375217470649c0e844955b -> 3c4b28b4d8072abd07405ffbe18c16de7c9d0a86 ([`pr/fatalresult.32`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.32) -> [`pr/fatalresult.33`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.33), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.32..pr/fatalresult.33))<!-- end --> to fix ci lint and tidy errors https://cirrus-ci.com/task/6460361353723904 https://cirrus-ci.com/task/5052986470170624
<!-- begin push-34 -->
Rebased 3c4b28b4d8072abd07405ffbe18c16de7c9d0a86 -> b1136354d6d7d195bfda2eb3b9bdf79f87cef3bf ([`pr/fatalresult.33`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.33) -> [`pr/fatalresult.34`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.34), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.33-rebase..pr/fatalresult.34))<!-- end -->
<!-- begin push-35 -->
Rebased b1136354d6d7d195bfda2eb3b9bdf79f87cef3bf -> aa57da149017205d4e0df201c8c301d520fc17d6 ([`pr/fatalresult.34`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.34) -> [`pr/fatalresult.35`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.35), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.34-rebase..pr/fatalresult.35))<!-- end --> due to conflict with #30595
<!-- begin push-36 -->
Rebased aa57da149017205d4e0df201c8c301d520fc17d6 -> 2bc7da5ec86c1174ade3e908b2844822019acf8f ([`pr/fatalresult.35`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.35) -> [`pr/fatalresult.36`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.36), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.35-rebase..pr/fatalresult.36))<!-- end -->
<!-- begin push-37 -->
Rebased 2bc7da5ec86c1174ade3e908b2844822019acf8f -> 15fa85163f324d50f6365b1865c320c974bcc3ee ([`pr/fatalresult.36`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.36) -> [`pr/fatalresult.37`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.37), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.36-rebase..pr/fatalresult.37))<!-- end -->
<!-- begin push-38 -->
Rebased 15fa85163f324d50f6365b1865c320c974bcc3ee -> f4e0a7d5eb59c6e74c2e3c2f0e7992251303975a ([`pr/fatalresult.37`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.37) -> [`pr/fatalresult.38`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.38), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.37-rebase..pr/fatalresult.38))<!-- end --> due to conflicts
<!-- begin push-39 -->
Updated f4e0a7d5eb59c6e74c2e3c2f0e7992251303975a -> 642e3a5b21b81317ad1a78a83456f0a2bbcd0fb9 ([`pr/fatalresult.38`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.38) -> [`pr/fatalresult.39`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.39), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.38..pr/fatalresult.39))<!-- end --> to fix CI errors https://github.com/bitcoin/bitcoin/actions/runs/21052613538
<!-- begin push-40 -->
Updated 642e3a5b21b81317ad1a78a83456f0a2bbcd0fb9 -> 871651378ca0af03365040f60858a2939e4bdf68 ([`pr/fatalresult.39`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.39) -> [`pr/fatalresult.40`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.40), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.39..pr/fatalresult.40))<!-- end --> to fix CI error https://github.com/bitcoin/bitcoin/actions/runs/21078196842/job/60625273966?pr=29700
<!-- begin push-41 -->
Rebased 871651378ca0af03365040f60858a2939e4bdf68 -> 75e142acf994773d7693a63e53193cd342dab78e ([`pr/fatalresult.40`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.40) -> [`pr/fatalresult.41`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.41), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.40-rebase..pr/fatalresult.41))<!-- end --> on top of #25665 pr/bresult2.73 due to conflicts with #34253 and #33680
<!-- begin push-42 -->
Updated 75e142acf994773d7693a63e53193cd342dab78e -> 326c23ab46e24fcf8114439d6e19f23ef1106968 ([`pr/fatalresult.41`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.41) -> [`pr/fatalresult.42`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.42), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.41..pr/fatalresult.42))<!-- end --> to fix IWYU errors https://github.com/bitcoin/bitcoin/actions/runs/21643714580/job/62390097850?pr=29700
<!-- begin push-43 -->
Rebased 326c23ab46e24fcf8114439d6e19f23ef1106968 -> ad8f98ceb224324d1d0fd81d45c83d04b1b3e92d ([`pr/fatalresult.42`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.42) -> [`pr/fatalresult.43`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.43), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.42-rebase..pr/fatalresult.43))<!-- end --> due to conflict with #34504
<!-- begin push-44 -->
Updated ad8f98ceb224324d1d0fd81d45c83d04b1b3e92d -> 09cbe90fa5c05e56d43895948d0d4f2b771cff9f ([`pr/fatalresult.43`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.43) -> [`pr/fatalresult.44`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.44), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.43..pr/fatalresult.44))<!-- end --> to fix feature_init.py failure https://github.com/bitcoin/bitcoin/actions/runs/21833788038/job/62998952469?pr=29700 caused by silent conflict with #34241
<!-- begin push-45 -->
Rebased 09cbe90fa5c05e56d43895948d0d4f2b771cff9f -> 8c5593b5110a32ee955bc5926d73aef74c5e99f6 ([`pr/fatalresult.44`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.44) -> [`pr/fatalresult.45`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.45), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.44-rebase..pr/fatalresult.45))<!-- end --> due to conflicts with #32950, #33512, #34184
<!-- begin push-46 -->
Rebased 8c5593b5110a32ee955bc5926d73aef74c5e99f6 -> 235dd2123953d8fff11abc5102d152635065d3f5 ([`pr/fatalresult.45`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.45) -> [`pr/fatalresult.46`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.46), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.45-rebase..pr/fatalresult.46))<!-- end --> on top of #25665 pr/bresult2.75 to fix CI failures https://github.com/bitcoin/bitcoin/actions/runs/22557596616/job/65337794687?pr=29700#step:9:862 caused by previous bad conflict resolutions
<!-- begin push-47 -->
Rebased 235dd2123953d8fff11abc5102d152635065d3f5 -> b3d85c7c6ad1fea59270532e78c02a2a3f4d40c1 ([`pr/fatalresult.46`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.46) -> [`pr/fatalresult.47`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.47), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.46-rebase..pr/fatalresult.47))<!-- end --> on top of #25665 pr/bresult2.76
<!-- begin push-48 -->
Rebased b3d85c7c6ad1fea59270532e78c02a2a3f4d40c1 -> 04122285bd2562b7edfa8eec59c22b2ef38e3b82 ([`pr/fatalresult.47`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.47) -> [`pr/fatalresult.48`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.48), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.47-rebase..pr/fatalresult.48))<!-- end --> on top of #25665 pr/bresult2.77
<!-- begin push-49 -->
Updated 04122285bd2562b7edfa8eec59c22b2ef38e3b82 -> efbb422a78f6e698e8aa3283ae1341aa6d8fbcaa ([`pr/fatalresult.48`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.48) -> [`pr/fatalresult.49`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.49), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.48..pr/fatalresult.49))<!-- end --> dropping commit that wasn't supposted to be part of PR, to fix -Wno-error=maybe-uninitialized false positive CI error https://github.com/bitcoin/bitcoin/actions/runs/23848113068/job/69520552767
<!-- begin push-50 -->
Rebased efbb422a78f6e698e8aa3283ae1341aa6d8fbcaa -> e5cbf121fcabb91a949b9d9c55f1a823e1c2aeaf ([`pr/fatalresult.49`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.49) -> [`pr/fatalresult.50`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.50), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.49-rebase..pr/fatalresult.50))<!-- end --> due to conflicts with #34208, #34858, and #34884
<!-- begin push-51 -->
Added 1 commits e5cbf121fcabb91a949b9d9c55f1a823e1c2aeaf -> 7868cc06ee51435e062462990e9b3fe256e72bc9 ([`pr/fatalresult.50`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.50) -> [`pr/fatalresult.51`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.51), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.50...pr/fatalresult.51))<!-- end --> to fix IWYU error https://github.com/bitcoin/bitcoin/actions/runs/24241244688/job/70776118438?pr=29700
<!-- begin push-52 -->
Rebased 7868cc06ee51435e062462990e9b3fe256e72bc9 -> 578d94d4c25e55b544f9f7ca2d58b09d399feeb2 ([`pr/fatalresult.51`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.51) -> [`pr/fatalresult.52`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.52), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.51-rebase..pr/fatalresult.52))<!-- end --> due to conflicts with #34435, #34440, #34440, and #34544 on top of #25665 pr/bresult2.77
<!-- begin push-53 -->
Rebased 578d94d4c25e55b544f9f7ca2d58b09d399feeb2 -> 82c181cb22c8826a9c68871a723dc7527c1c8532 ([`pr/fatalresult.52`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.52) -> [`pr/fatalresult.53`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.53), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.52-rebase..pr/fatalresult.53))<!-- end --> on top of #25665 pr/bresult2.79 due to conflict with #34644
<!-- begin push-54 -->
Updated 82c181cb22c8826a9c68871a723dc7527c1c8532 -> 27a3713657c497c9cc5de7c7dcdfc4c9c2b4ece8 ([`pr/fatalresult.53`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.53) -> [`pr/fatalresult.54`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.54), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.53..pr/fatalresult.54))<!-- end --> to fix iwyu missing kernel/result.h and clang-tidy unused AbortFailure decl https://github.com/bitcoin/bitcoin/actions/runs/27140938629/job/80106968297
<!-- begin push-55 -->
Updated 27a3713657c497c9cc5de7c7dcdfc4c9c2b4ece8 -> 84f625c950fd46f4ac09f2b290973f14e25e4443 ([`pr/fatalresult.54`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.54) -> [`pr/fatalresult.55`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.55), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.54..pr/fatalresult.55))<!-- end --> to fix iwyu errors: missing #include <kernel/result.h> in bench/connectblock.cpp and #include <tuple> in bench/block_assemble.cpp https://github.com/bitcoin/bitcoin/actions/runs/27152433227/job/80146651632
<!-- begin push-56 -->
Rebased 84f625c950fd46f4ac09f2b290973f14e25e4443 -> 3ff7b8f8480b27bd55ea959b8fb7182b11b829e4 ([`pr/fatalresult.55`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.55) -> [`pr/fatalresult.56`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.56), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.55-rebase..pr/fatalresult.56))<!-- end --> on top of #25665 pr/bresult2.80 due to conflicts with #35182 and #35359
<!-- begin push-57 -->
Updated 3ff7b8f8480b27bd55ea959b8fb7182b11b829e4 -> 5c2680891e53daa7a92423df8f76ee42f0a4ade9 ([`pr/fatalresult.56`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.56) -> [`pr/fatalresult.57`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.57), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.56..pr/fatalresult.57))<!-- end --> to fix iwyu errors: add missing includes to bench/load_external.cpp and bench/readwriteblock.cpp https://github.com/bitcoin/bitcoin/actions/runs/28116002760/job/83255512311
<!-- begin push-58 -->
Updated 5c2680891e53daa7a92423df8f76ee42f0a4ade9 -> a07e8a22295289c5b57c62c8fec37fa75457cf91 ([`pr/fatalresult.57`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.57) -> [`pr/fatalresult.58`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.58), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.57..pr/fatalresult.58))<!-- end --> updating comments to address https://github.com/bitcoin/bitcoin/pull/29700#discussion_r3446525045 and clarify the different ways flush errors are ignored
<!-- begin push-59 -->
Rebased a07e8a22295289c5b57c62c8fec37fa75457cf91 -> 03f2c81adc1917d0c1e292d2287d7868bf424949 ([`pr/fatalresult.58`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.58) -> [`pr/fatalresult.59`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.59), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.58-rebase..pr/fatalresult.59))<!-- end --> due to conflicts with #35295, #35621, #35652, and #35673
<!-- begin push-60 -->
Rebased 03f2c81adc1917d0c1e292d2287d7868bf424949 -> 6542bfe144b1f735d628ce59973c37a55a48a7fc ([`pr/fatalresult.59`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.59) -> [`pr/fatalresult.60`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.60), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.59-rebase..pr/fatalresult.60))<!-- end --> due to conflict with #35572
<!-- begin push-61 -->
Rebased 6542bfe144b1f735d628ce59973c37a55a48a7fc -> b6d2a6097af5f2592a4b58dca264119670d9f3ef ([`pr/fatalresult.60`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.60) -> [`pr/fatalresult.61`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.61), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.60-rebase..pr/fatalresult.61))<!-- end --> on top of #25665 pr/bresult2.82 due to conflicts with #35847 and #35205
<!-- begin push-62 -->
Rebased b6d2a6097af5f2592a4b58dca264119670d9f3ef -> bef716036337413bd277cf8b8d2c3527f752a873 ([`pr/fatalresult.61`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.61) -> [`pr/fatalresult.62`](https://github.com/ryanofsky/bitcoin/commits/pr/fatalresult.62), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fatalresult.61-rebase..pr/fatalresult.62))<!-- end --> updating blockfilter_index_tests.cpp for new ProcessNewBlock signature after #35847 https://github.com/bitcoin/bitcoin/actions/runs/31894652860/job/95035991561?pr=29700, on top of #25665 pr/bresult2.82

ryanofsky committed:
ci: Avoid -Wno-error=maybe-uninitialized false positives

Avoid false positive maybe-uninitialized errors in native_previous_releases
CI job.

Fix was suggested by maflcko in
https://github.com/bitcoin/bitcoin/pull/25665#issuecomment-3457573218

CI errors look like:
https://github.com/bitcoin/bitcoin/actions/runs/25426711469/job/74581986567?pr=25665

/home/admin/actions-runner/_work/_temp/src/wallet/test/util.cpp: In function ‘wallet::DescriptorScriptPubKeyMan* wallet::CreateDescriptor(CWallet&, const std::string&, bool)’:
/home/admin/actions-runner/_work/_temp/src/wallet/test/util.cpp:143:10: error: ‘*(const std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>*)((char*)&spkm + offsetof(util::Result<std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>, void, util::Messages>,util::Result<std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>, void, util::Messages>::<unnamed>.util::detail::SuccessHolder<std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>, void, util::Messages>::<unnamed>)).std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>::_M_data’ may be used uninitialized [-Werror=maybe-uninitialized]
  143 |     auto spkm = Assert(keystore.AddWalletDescriptor(w_desc, keys,/*label=*/"", /*internal=*/false));
      |          ^~~~
cc1plus: all warnings being treated as errors

optout21 commented on src/validation.cpp:
```diff
@@ -4378,14 +4462,15 @@ bool ChainstateManager::AcceptBlock(const std::shared_ptr<const CBlock>& pblock,
     // background validation yet).
-    ActiveChainstate().FlushStateToDisk(state, FlushStateMode::NONE);
+    // Ignore failure value, do not treat flush error as failure.
+    ActiveChainstate().FlushStateToDisk(state, FlushStateMode::NONE) >> result;
```
While working on #35570 , I have also found a potential inconsistency here: the return value from `FlushStateToDisk` is ignored, but it can cause change in `state` -- unclear if this was intended or just an unintended side effect.
See https://github.com/bitcoin/bitcoin/pull/35570#issuecomment-4758227624

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29700#discussion_r3446525045

> While working on #35570 , I have also found a potential inconsistency here: the return value from `FlushStateToDisk` is ignored, but it can cause change in `state` -- unclear if this was intended or just an unintended side effect. See [#35570 (comment)](https://github.com/bitcoin/bitcoin/pull/35570#issuecomment-4758227624)

Yes thank you for pointing out this inconsistency. It existed before this PR and this PR doesn't change it but the PR does add a comment and it's better if the comment acknowledges it. Updated in latest push to clarify this comment and related ones.

To describe the issue for others: validation code flushes internally many different places and mostly does not treat flush failures as errors returned to callers. This PR doesn't change that behavior but it does return additional information to callers about whether flushes happened and whether they failed. Most of the validation code that flushes (examples: AcceptToMemoryPool, ConnectTip, ActivateBestChainStep) pass a `state_dummy` variable to the flush code to keep flush errors separate from normal errors. But this code is an exception and passes `state` instead so it's possible for this function to return true even though `state.IsValid()` is false, and for `LoadExternalBlockFile` to fail when it might succeed otherwise. This inconsistency should probably be fixed one way or another, but it's a little outside the scope of this PR so the PR is just adding a comment

optout21 replied:
Addressing this single case has been split off into #35621 .

ryanofsky committed:
Merge branch 'pr/bresult2' into pr/fatalresult

ryanofsky committed:
test-each-commit: Increase fetch depth

Needed due to base PR, can be dropped before merge
